Capture vars not env

chrish42 opened the pull request:
This is the current state of my work  to capture only the required variables for patsy formulas instead of the whole environment.

Current status: 
- [x] Add new methods to EvalEnvironment to create a environment that contains only a subset of variables.
- [x] Refactor EvalFactor (and update all users) to move the eval_env parameter out of the its initializer method and into the state.
- [x] Additions to doc/changes.rst (esp. to document the incompatible changes!)
- [x] Do a quick pass through the docs to catch anything that's now out-of-date.
- [x] Add "end-to-end" test or two -- something that tests the user-visible behaviour directly, like running a formula through dmatrix, then modifying the original environment, and then running some more data through the same builder (like we were doing predictions) and checking that it is unaffected by our changes to the env. (test_highlevel.py is where the other tests like this go.)
- [x] Create right subset EvalEnvironment and stash it into the state dict for EvalFactor.
- [x] ~~Figure out what the right thing to do is when variables are shadowed between the EvalFactor's environment and the data.~~ (Can be done together with #13 instead. Not a blocker for this.)

When this is complete, it will fix bug #25.

njsmith commented on patsy/build.py:
```diff
@@ -642,14 +650,25 @@ def design_matrix_builders(termlists, data_iter_maker, NA_action="drop"):
+    # more obscure backtrace later on.
+    if not isinstance(eval_env, six.integer_types + (EvalEnvironment,)):
+        raise TypeError("Parameter 'eval_env' must be either an integer or an instance" +
+                        "of patsy.EvalEnvironment.")
```
Nitpicks:
* The "+" is unnecessary here -- the Python lexer (like that in many languages, e.g. C) will concatenate adjacent string tokens at compile time, e.g. `"foo" "bar"` is the same as `"foobar"` (and so is `"foo" 'bar'` for that matter). It's even slightly more efficient, b/c `+` is evaluated at runtime, while leaving it out gives a single string directly in the .pyc. Not that this efficiency matters at all :-). But as a style/consistency thing I leave the `+` out.
* You're missing a space, this string says "instanceof" :-)

chrish42 replied:
Yeah. I'm aware of this feature, but I vaguely remember being bitten by it (although maybe it was in C instead of Python), so I personally tend to prefer being explicit unless necessary. But given that this is the style of the rest of the code base, I'll go with it. Done.

njsmith commented on patsy/desc.py:
```diff
@@ -593,8 +592,7 @@ def _assert_terms_match(terms, expected_intercept, expecteds, eval_env): # pragm
                 expected = (expected,)
-            assert term.factors == tuple([EvalFactor(s, eval_env)
-                                          for s in expected])
+            assert term.factors == tuple([EvalFactor(s) for s in expected])
```
I think `eval_env` can probably be removed more thoroughly from this code? (I.e., if there aren't any other users in the function, then the variable can be removed entirely, and so on up the call stack.)

chrish42 replied:
It didn't percolate very far off the call stack, but I removed a couple uses. Good catch.

njsmith commented on patsy/eval.py:
```diff
@@ -81,6 +82,42 @@ def test_VarLookupDict():
+
+    :arg code: A string containing a Python expression.
+    """
+    disallowed_ast_nodes = (ast.Lambda, ast.ListComp, ast.GeneratorExp)
```
Let's add a comment here saying something like "# Syntax that allows new name bindings to be introduced is tricky to handle here, so we just refuse to do so"

chrish42 replied:
Good point. Done.

njsmith commented on patsy/eval.py:
```diff
@@ -81,6 +82,42 @@ def test_VarLookupDict():
+
+    for node in ast.walk(ast.parse(code)):
+        if isinstance(node, disallowed_ast_nodes):
+            raise PatsyError("Lambda, list/dict/set comprehension, generator expression in patsy formula not currently supported.")
```
Please break this line up -- we try to stick below 80 characters per line.

chrish42 replied:
Done.

njsmith commented on patsy/eval.py:
```diff
@@ -460,16 +522,16 @@ def new_name_maker(token):
     def memorize_chunk(self, state, which_pass, data):
         for obj_name in state["pass_bins"][which_pass]:
-            self._eval(state["memorize_code"][obj_name], state, data)
+            self._eval(state["memorize_code"][obj_name], state["eval_env"],state, data)
```
Missing space after comma, and probably a long line as well?

chrish42 replied:
Added space, put each arg on its own line.

njsmith commented on patsy/highlevel.py:
```diff
@@ -33,7 +33,7 @@
                        NA_action):
     if isinstance(formula_like, DesignMatrixBuilder):
-        return (design_matrix_builders([[]], data_iter_maker, NA_action)[0],
+        return (design_matrix_builders([[]], data_iter_maker, eval_env, NA_action)[0],
```
long line?

chrish42 replied:
86 chars, of which I've seen a couple other ones in the code base. Would you rather have me leave it as is, or split the arguments of the `design_matrix_builders()` function call each on its own line?

njsmith replied:
Eh, whatever :-)

njsmith commented on patsy/highlevel.py:
```diff
@@ -53,7 +53,7 @@ def _try_incr_builders(formula_like, data_iter_maker, eval_env,
         return design_matrix_builders([formula_like.lhs_termlist,
                                        formula_like.rhs_termlist],
-                                      data_iter_maker,
+                                      data_iter_maker, eval_env,
```
Another trivial nitpick: I prefer to either put all args on one line, or else put each arg on its own line -- easier to read that way IMO

chrish42 replied:
Oky doke. Done

njsmith commented:
Hi Christian! As you can see I've mostly recovered from PyCon, and did a quick pass for general style on what you have so far :-). Looking forward to seeing the rest!

njsmith commented on patsy/build.py:
```diff
@@ -614,7 +614,7 @@ def _make_term_column_builders(terms,
     return new_term_order, term_to_column_builders
 
-def design_matrix_builders(termlists, data_iter_maker, NA_action="drop"):
+def design_matrix_builders(termlists, data_iter_maker, eval_env=0, NA_action="drop"):
```
After thinking about this a little more... I think maybe we should actually make `eval_env` a required argument here. This is the lower-level machinery for those who are doing clever programmatic things, so it's safer I think to insist the user think about how they want to handle the evaluation environment.

chrish42 replied:
Sounds good to me too. Done.

chrish42 commented:
Hi Nathaniel! Thanks for the review. I will reply to them / apply them once I'm done with fixing the tests for using a subset of the environment. 

I am currently in the process of fixing said tests. Instead of comparing `state["eval_env"]` to the captured EvalEnvironment, I now build an EvalEnvironment containing the correct subset, and compare it to `state["eval_env"]`. However, that cannot work, as the comparison for equality of EvalEnvironments only compares on identity of each namespace dict. How would you feel if I made `EvalEnvironment.__eq__` check that each namespace dict has the same keys (by equality) and values (by identity) instead?

njsmith commented:
Hi Christian,

Sounds like a good plan. Before I answer your actual question, can you back
up a step and elaborate on why you need equality comparisons in the first
place? The old code needed to define equality on EvalEnvironment because it
had to define equality on EvalFactor, and EvalFactor used to have an
EvalEnvironment. But now that it doesn't, I think EvalEnvironment.__eq__
should just be unused (and maybe even removed)?
On Apr 29, 2015 9:50 AM, "Christian Hudon" <notifications@github.com> wrote:

> Hi Nathaniel! Thanks for the review. I will reply to them / apply them
> once I'm done with fixing the tests for using a subset of the environment.
>
> I am currently in the process of fixing said tests. Instead of comparing
> state["eval_env"] to the captured EvalEnvironment, I now build an
> EvalEnvironment containing the correct subset, and compare it to
> state["eval_env"]. However, that cannot work, as the comparison for
> equality of EvalEnvironments only compares on identity of each namespace
> dict. How would you feel if I made EvalEnvironment.__eq__ check that each
> namespace dict has the same keys (by equality) and values (by identity)
> instead?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pydata/patsy/pull/66#issuecomment-97502065>.
>

chrish42 commented:
I need it for the tests. Here is one of the tests that I am in the process of fixing up (in eval.py):

```python
def test_EvalFactor_memorize_passes_needed():
    from patsy.state import stateful_transform
    foo = stateful_transform(lambda: "FOO-OBJ")
    bar = stateful_transform(lambda: "BAR-OBJ")
    quux = stateful_transform(lambda: "QUUX-OBJ")
    e = EvalFactor("foo(x) + bar(foo(y)) + quux(z, w)")

    state = {}
    eval_env = EvalEnvironment.capture(0)
    passes = e.memorize_passes_needed(state, eval_env)
    expected_eval_env = EvalEnvironment([{'foo': foo,
                                                  'bar': bar,
                                                  'quux': quux}],
                                                  eval_env.flags)
    assert passes == 2
    assert state["eval_env"] == expected_eval_env
# More stuff cut.
```

The last line is why I need equality on EvalEnvironment. (Before, the code just did `assert state["eval_env"] == eval_env`. Maybe my previous comment now makes more sense in light of this code? I think equality on EvalEnvironment needs to be defined (likely) as I described it for these tests. Although, if equality is used only for the tests, we could also pull it out of the class and put it into a test-only helper function? Anyways, let me know what you think is best here.

njsmith commented:
Ah, I see. My preference would still be to not define equality on `EvalEnvironment` at all (just because semantically it's extremely tricky -- e.g. two `EvalEnvironment` objects might act differently if they have different `dict` objects in them, because one thing you can do with an `EvalEnvironment` is execute code that modifies stuff in the environment -- and we don't really need it.) And for tests like this, just test that the `EvalEnvironment` does in fact have the correct behavior, e.g.:

```
# Variables in the current environment that were mentioned in the formula are accessible
assert state["eval_env"].eval("foo") is foo
# But other variables in the current environment that were not mentioned, are not available
assert_raises(NameError, state["eval_env"].eval, "state")
```

(This isn't as thorough as a full equality check, but we already have thorough tests for `EvalEnvironment.subset` and `ast_names`, right? So it isn't crucial to reproduce those tests here IMHO.)

chrish42 commented:
Sounds reasonable to me. I'll move in that direction. I'll update the pull request when all the tests pass again.

chrish42 commented:
Here's the updated pull request, using the subset of the environment in the state. Let me know what you think.

chrish42 commented:
Also, what do you think is the right thing to do regarding shadowing of variables between the EvalEnvironment and the DataFrame or other passed to the formula?

njsmith commented:
Looking pretty good! Things I can think of besides the shadowing issue:
* we need some additions to `doc/changes.rst` (esp. to document the incompatible changes!)
* need to do a quick pass through the docs to catch anything that's now out-of-date.
* I'd like to see an "end-to-end" test or two -- something that tests the user-visible behaviour directly, like running a formula through `dmatrix`, then modifying the original environment, and then running some more data through the same builder (like we were doing predictions) and checking that it is unaffected by our changes to the env. (test_highlevel.py is where the other tests like this go.)

njsmith commented:
Okay, so having had a chance to look over the code that would be affected by shadowing, I think my thought is: some sort of systematic handling of shadowing would probably be a good idea, but it feels like something we can just as well handle as an independent change. In particular, shadowing checks should probably also notice when builtins are shadowed (this has been a persistent issue -- see #13). So I don't think it's a blocker for this PR.

chrish42 commented:
Working on the doc now. I assume you will register this release on zenodo before release time? (I'm also assuming the next release will be v0.4.0.)

njsmith commented:
Actually, an annoying thing about Zenodo is that you actually cannot get the DOI until after the release is finalized, basically b/c it the DOI-generation operation takes the release hash as an input. So I've just been putting the DOIs into the dev docs after the release, and the dev docs are what show up by default on readthedocs anyway, and I guess it's good enough.

And yeah, we'll call it v0.4.0.

chrish42 commented:
Just updated the pull request with the results of my pass over the docs. Could you go over the diff for the doc update and see if things still make sense? I've deleted a lot more eval_env, etc. stuff than I added to the docs. I'm a bit worried that this means we missed an additional place where an eval_env argument should have been added.

I've also mentioned the incompatible change to `EvalFactor.memorize_passes_needed()` in the changelog, even though this is a pretty low-level method (which doesn't have a docstring too). Not too much detail for changes.rst? Anyways, feedback on my doc changes just as welcome as on the code ones.

I'll add the high-level tests another day, but other than that, this is starting to look pretty complete.

njsmith commented:
The doc changes look good to me -- I think the docs just spend more time talking about the factor interface than the actual builder interface, so it makes sense that you deleted more than you added. Mentioning the `EvalFactor.memorize_passes_needed` change is fine -- probably no-one cares, but oh well, it won't hurt.

chrish42 commented:
Added a high-level test. Let me know if you have other ideas for these kinds of tests. Otherwise, please let me know if you have any final comments, if there are still missing bits, etc. (Really looking forward to getting this merged.)

njsmith commented:
Looks good to me!

One last thing: it looks like you have some merge conflicts which are greying out the merge button and preventing the tests from running. Could you merge or rebase from current master to resolve these?

chrish42 commented:
Merge done.

njsmith commented:
Huzzah! We are good to go.

njsmith commented:
Thanks for the great work, and for sticking with it despite my nitpickiness :-).

chrish42 commented:
Woot! Great!

chrish42 commented:
Haha. No problem! Happy to be starting to make contributions on the scientific Python software stack. More of your comments were very relevant, I thought. And until I come close to you in terms of code contributed to patsy, it's much more your code base than mine, so some nitpickiness is fully allowed. (I tried already to stick to your coding style as best as I could.) I also much prefer code bases with a unified style anyway, even when I don't agree 100% with it. So no problem again.
